### ISSUE TITLE
ARP responses are not sent when the network interface is a bridge

### ISSUE THREAD

**Issue by johnlane** [metallb/metallb#692]
I am running a single node cluster on a development machine using k3s with metallb. I first approached #metallb but got no response. Having spent two weeks trying to solve this, I hope this is a valid bug report.

If the node's network interface is bridged then MetalLB does not respond to ARP requests. The requests arrive, confirmed with Wireshark, but no response is sent. If the node's network interface is not bridged then MetalLB does respond to ARP requests as expected.

Adding a static entry to a client's ARP table allows it to access the service, so it's only the ARP response that doesn't work. I can also fudge an ARP response by adding the load-balanced IP to the node's loopback interface.

Other bridged services (e.g. kvm instances ) work as they should.

I am running k3s version v1.17.4+k3s1 (3eee8ac3) with MetalLB 0.9.3 installed using the apply method (not Helm). I'm using nginx ingress controller and coredns as supplied by k3s. Apart from replacing traefik/servicelb with nginx/metallb and using docker backend the installation is vanilla. I am running on Arch Linux (currently kernel 5.6.11).

```
$ kubectl describe configmap config --namespace metallb-system
Name:         config
Namespace:    metallb-system
Labels:       <none>
Annotations:
Data
====
config:
----
address-pools:
- name: default
  protocol: layer2
  addresses:
  - 192.168.21.240-192.168.21.245

Events:  <none>
```
To reproduce this you can...

```
$ kubectl apply -f nginx-hello-loadbalancer.yaml
$ kubectl get service -o wide
hello        LoadBalancer   10.43.101.235   192.168.21.242   80:30728/TCP   24m   app=hello 
```
Then, from another machine on the same subnet, browse to `http://192.168.21.242` and observe ARP on the k3s node (e.g. using Wireshark):
```
30	3.150454914	IntelCor_03:02:f0	Broadcast	ARP	60	Who has 192.168.21.242? Tell 192.168.21.221
```
(many times, no response).

Stop the network and start it without the bridge. Refresh the browser. It works.

Put the bridge back. Clear the client's arp table entry ( `arp -d 192.168.21.242`). Refresh the browser. It doesn't work.

Either add static arp on client (`arp -s 192,168,21,242 MACADDR`) or add IP on host loopback (`ip address add 192.168.21.242/32 dev lo`). Refresh the browser. It works.

**Comment by rata** [comment]
@johnlane thanks for the bug report. Can you share how you configure the bridge? When using `kind` locally with a simple bridge, I can see the ARP response.

If you sniff the traffic in the MetalLB speaker pod, do you see the package arriving there?

If you can provide more clear instructions on how to reproduce (kind would be great, as is very easy to run locally), it would help a lot.


Thanks!

**Comment by johnlane** [comment]
@rata sorry was away for a few days...

I'm using k3s because I had problems with kind due to my system running on zfs. I'm using the archlinux `netctl` network scripts but I have also created a bridge manually using the same sequence of commands those scripts use:

    ip link add name br0 type bridge
    ip link set dev enp0s31f6 promisc on up
    ip addr flush dev enp0s31f6 scope host
    ip addr flush dev enp0s31f6 scope site
    ip addr flush dev enp0s31f6 scope global
    ip link set dev enp0s31f6 master br0
    dhcpcd -4 -q -t 30 -L br0

I don't know if all those commands are strictly necessary but that is what the arch script does, and doing same by hand also results in a bridge where the metallb arp does not respond. Here's the details of the bridge created

    # ip address show dev br0
    78: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
        link/ether 32:58:84:2d:c8:33 brd ff:ff:ff:ff:ff:ff
        inet 192.168.21.236/24 brd 192.168.21.255 scope global dynamic noprefixroute br0
           valid_lft 86327sec preferred_lft 75527sec
        inet6 fe80::3058:84ff:fe2d:c833/64 scope link
           valid_lft forever preferred_lft forever

    # ip address show dev enp0s31f6
    2: enp0s31f6: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br0 state UP group 
    default qlen 1000
        link/ether e4:b9:7a:35:ef:af brd ff:ff:ff:ff:ff:ff
        inet6 fe80::e6b9:7aff:fe35:efaf/64 scope link
           valid_lft forever preferred_lft forever

Re trying to sniff the traffic in the speaker pod, I'm not familiar with doing that but I added a "sidecar" container to the speaker pod containing tcpdump and used that to capture packets (using [this](https://developers.redhat.com/blog/2019/02/27/sidecars-analyze-debug-network-traffic-kubernetes-pod) and [this](https://www.techrepublic.com/article/how-to-patch-a-running-kubernetes-pod) for inspiration).

So I'm doing this inside that container within the speaker pod (that seems to just have the host's network stack - is that right?):

    $ tcpdump -n -i any arp
    13:24:50.358354 ARP, Request who-has 192.168.21.240 tell 192.168.21.221, length 46
    13:24:50.358487 ARP, Reply 192.168.21.240 is-at e4:b9:7a:35:ef:af, length 46

That's what I see without the bridge. If I add the bridge then I only see the request, repeated multiple times, but no response.

    $ tcpdump -n -i any arp
    13:30:58.973690 ARP, Request who-has 192.168.21.240 tell 192.168.21.221, length 46
    13:30:58.973690 ARP, Request who-has 192.168.21.240 tell 192.168.21.221, length 46

I'm going to do some more testing, and also see if I can reproduce using `kind` for you. I just wanted to push this info as soon as I could...

**Comment by johnlane** [comment]
Hello, another update. I updated my `kind` to 0.9.0 yesterday in order to test with that. It worked, but I don't think it's an apples-to-apples comparison because you're on a host-local network that's private to the host and the `kind` containers as far as I can tell. So the real bridge isn't party to the action.

But the wierd thing is, I tried `k3s` again, not aware of having changed anything, and it is now working. I have no idea why at the moment (and I hate when things start working without a reason, even though having them work is great!). I'll report back any differences to this statement.

**Comment by johnlane** [comment]
Hello, another update... I just did a complete reinstall of my laptop and have metalb 0.95 on k3s. The problem is back (not sure it actually went away). However, I can fix it. I believe it is to do with speaker failing on createNDPResponder and then not bothering to set up ARP on that interface. I discovered [this issue](https://github.com/metallb/metallb/issues/248#issuecomment-634801881) which led me down the path to a solution.

In the speaker logs, I see loads of this:
```
{"caller":"announcer.go:114","error":"creating NDP responder for \"br-c6c0203cdd09\": listen ip6:ipv6-icmp fe80::1%br-c6c0203cdd09: bind: cannot assign requested address","interface":"br-c6c0203cdd09","msg":"failed to create NDP responder","op":"createNDPResponder","ts":"2020-11-24T09:25:03.273205182Z"}
```
I disabled ipv6 on that interface, like this:
```
echo 1 > /proc/sys/net/ipv6/conf/br-c6c0203cdd09/disable_ipv6
```
And, immediately, the speaker log showed
```
{"caller":"announcer.go:109","event":"createARPResponder","interface":"br-ce87a5a2060c","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.283462889Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"br-cf685aea22ca","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.283636842Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"br-17bbbb53d7f9","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.283782623Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"docker0","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.283925175Z"}
{"caller":"announcer.go:118","event":"createNDPResponder","interface":"docker0","msg":"created NDP responder for interface","ts":"2020-11-24T09:26:23.284107719Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"cni0","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.284530583Z"}
{"caller":"announcer.go:118","event":"createNDPResponder","interface":"cni0","msg":"created NDP responder for interface","ts":"2020-11-24T09:26:23.284618824Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"br0","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.284909695Z"}
{"caller":"announcer.go:109","event":"createARPResponder","interface":"flannel.1","msg":"created ARP responder for interface","ts":"2020-11-24T09:26:23.285062524Z"}
{"caller":"announcer.go:118","event":"createNDPResponder","interface":"flannel.1","msg":"created NDP responder for interface","ts":"2020-11-24T09:26:23.285149043Z"}
```
and the arp requests are now responded to correctly (I tested a bridged vm on my laptop and also another physical machine).

So it would appear that speaker gives up on the first error and doesn't bother trying further interfaces.

Some more information about the problem interface:
```
$ ip address show dev br-c6c0203cdd09
5: br-c6c0203cdd09: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:f5:9c:b7:ea brd ff:ff:ff:ff:ff:ff
    inet 192.168.96.1/20 brd 192.168.111.255 scope global br-c6c0203cdd09
       valid_lft forever preferred_lft forever

$ docker network inspect c6c0203cdd09
[
    {
        "Name": "kind",
        "Id": "c6c0203cdd09a050353a19de582ec8e6f958f9e73797eaa7e12120c5652ee266",
        "Created": "2020-07-08T12:18:35.297660335+01:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": true,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "192.168.96.0/20",
                    "Gateway": "192.168.96.1"
                },
                {
                    "Subnet": "fc00:f853:ccd:e793::/64",
                    "Gateway": "fc00:f853:ccd:e793::1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```
I think that is a left-over from an earlier time, when I was trying kind. I've done a clean-up now, reset Docker and cleaned out stale stuff. Everything now appears to be working as it should.

I think what falls out of this issue is the fact that the speaker gives up on the first error it encounters.

**Comment by JuniorJPDJ** [comment]
I'm having the same issue. Creating veth and adding one side to the bridge and second side to the metallb doesn't solve the issue.

**Comment by JuniorJPDJ** [comment]
bump

**Comment by fedepaol** [comment]
> `arping` works (metallb sends ARP responses) with veth, but I cannot connect to the service

This means the issue is different from the current one, and that metallb did its part correctly. Curling from the node doesn't involve MetalLB.

Can you describe your topology? Is the client on the same l2 network? I am wondering if you are hitting an asymmetric return path and the traffic is being blocked by RPF.

**Comment by johnlane** [comment]
I think the issue that needs to be tracked here is the fact that the speaker gives up on the first error and doesn't bother trying further interfaces. Anything else should be a new issue?

It would be interesting to learn whether the original issue (speaker giving up on the first error without trying further interfaces) as been corroborated and is at least tracked for a fix.
